prov-api/helpers/prov_connector.py

```diff
@@ -0,0 +1,1597 @@
+        lineage_paths=paths
+    )
+
+    # Return D3.js friendly serialisation
+    return networkx.node_link_data(graph)
+
+
+def downstream_query(starting_id: str, depth: int, config: Config) -> Dict[str, Any]:
+    """    lineage_query
+        Performs a depth limited downstream query for a neo4j database. 
+        This query is defined as, starting at a node identified uniquely by
+        a handle id, traverse any relation in the child direction i.e. 
+        (starting_node)<-[]-(child_node).
+
+        The multiplicity of this exploration step is bounded by the depth 
+        provided. 
+
+        The result from the neo4j query is a list of records. A record is a kv
+        store of returned elements from the cypher query. This is parsed into a 
+        list of path objects. A path is a start, end and a list of steps/relationships
+        along the way. 
+
+        These paths are explored to generate a graph. The graph is a Networkx.DiGraph. 
+        This digraph can be returned in a D3.js friendly json serialisation format.        
+
+        Arguments
+        ----------
+        starting_id : str
+            The handle ID which should uniquely identify a node in the graph
+        depth : int
+            The maximum depth to traverse in the <-[*1..depth]- wildcard 
+        config : Config
+            Connection config etc
+
+        Returns
+        -------
+         : Dict[str, Any]
+            The DiGraph serialisation format which is D3.js friendly.
+            This format is JSON and contains a list of nodes with an id, 
+            category and subtype + a list of links which have a from, 
+            to and a relationship type.
+
+        See Also (optional)
+        --------
+
+        Examples (optional)
+        --------
+    """
+    if config.mock_graph_db:
+        raise RuntimeError(
+            "Asking for real data from graph DB during mock! Returning [].")
+
+    # The cypher query to make for lineage
+    query = f"""
+    MATCH r=((start{{`{IDENTIFIER_TAG}`:'{starting_id}'}})<-[*1..{depth}]-(child)) RETURN r
+    """
+
+    # Run the query
+    try:
+        result = run_query(query, config)
+    except Exception as e:
+        raise HTTPException(
+            status_code=500,
+            detail=f"Neo4j lineage query experienced an exception: {e}."
+        )
+
+    # Convert to list of paths by pulling out the 'r' field from the RETURN r in above
+    # cypher query
+    paths: List[Optional[graph.Path]] = [record.get('r') for record in result]
+
+    # Apply simple exploration algorithm to traverse paths and produce networkx digraph
+    graph = generate_networkx_graph_from_lineage_paths(
+        lineage_paths=paths
+    )
+
+    # Return D3.js friendly serialisation
+    return networkx.node_link_data(graph)
+
+
+class NodeLinkProps(BaseModel):
+    """
+    Represents properties of a link between nodes in the graph.
+    """
+
+    # Comma separated list of IDs (required)
+    record_ids: str
```

Comment:
validator to ensure it is a correctly formatted csv?

Reply:
I don't think it needs/should have a validator since a CSV list of IDs can be pretty much any string